### PR TITLE
Ade 613 render macro with velocity template

### PR DESCRIPTION
**Pull request by shouvSarker**

### CODE REVIEW

**Review by robinMcA — APPROVED** [review]
But give Dan a chance to look at it

**Comment by shouvSarker** [comment]
>  You can't instantiate a new entry in the `map`?

Updated to 
`    Map<String, String> renderedData = data.entrySet()
      .stream().map(entry ->
        new String[]{WordUtils.capitalize(entry.getKey().replaceAll("[A-Z]", " $0")), entry.getValue()}
      )
      .collect(Collectors.toMap(entry -> entry[0],entry->entry[1]));`

**Comment by shouvSarker** [comment]
> You can capitalise in a second step (as you already do). I don't think your splitting step is adequate (e.g. what should it do with safeJSON?) and is meaningless to reproduce when you likely have commons in the classpath already.

Used `splitByCharacterTypeCamelCase` to split on capital letters and then added capitalisation as a second step. Carved it to its own private function.

File: backend/src/main/java/au/com/agiledigital/idea_search/macros/StructuredData.java [hunk 1]
```diff
@@ -1,62 +1,51 @@
...
-  private static final Logger log = LoggerFactory.getLogger(StructuredData.class);
+  private Gson gson = new Gson();
 
   @Override
-  public String execute(Map<String, String> map, String s, ConversionContext conversionContext)
-    throws MacroExecutionException {
+  public String execute(Map<String, String> map, String s, ConversionContext conversionContext) {
+    Map<String, String> data = new LinkedHashMap<>();
+    // gets the page body data as mapped
+    data =gson.fromJson(Jsoup.parse(s).body().text(), data.getClass());
```
Comment (dspasojevic) [inline backend/src/main/java/au/com/agiledigital/idea_search/macros/StructuredData.java]: Why initialise `data` then re-assign?

File: backend/src/main/java/au/com/agiledigital/idea_search/macros/StructuredData.java [hunk 2]
```diff
@@ -1,62 +1,51 @@
...
+
+    // making keys (section headers) capitalised. splitting on capital letters, i.e. camelCase becomes Camel Case
+    Map<String, String> renderedData = data.entrySet()
+      .stream().map(entry ->
+        new String[]{ headingTransformation(entry.getKey()), entry.getValue()}
+      )
+      .collect(Collectors.toMap(entry -> entry[0],entry->entry[1],
```
Comment (dspasojevic) [inline backend/src/main/java/au/com/agiledigital/idea_search/macros/StructuredData.java]: Is this formatted?

File: backend/src/main/java/au/com/agiledigital/idea_search/macros/StructuredData.java [hunk 3]
```diff
@@ -1,62 +1,51 @@
...
-    throws MacroExecutionException {
+  public String execute(Map<String, String> map, String s, ConversionContext conversionContext) {
+    Map<String, String> data = new LinkedHashMap<>();
+    // gets the page body data as mapped
+    data =gson.fromJson(Jsoup.parse(s).body().text(), data.getClass());
+
+    // making keys (section headers) capitalised. splitting on capital letters, i.e. camelCase becomes Camel Case
+    Map<String, String> renderedData = data.entrySet()
+      .stream().map(entry ->
+        new String[]{ headingTransformation(entry.getKey()), entry.getValue()}
```
Comment (dspasojevic) [inline backend/src/main/java/au/com/agiledigital/idea_search/macros/StructuredData.java]: Is possible to instantiate an `Entry` in the version of java that is used?

File: backend/src/main/java/au/com/agiledigital/idea_search/macros/StructuredData.java [hunk 4]
```diff
@@ -1,62 +1,51 @@
...
+      )
+      .collect(Collectors.toMap(entry -> entry[0],entry->entry[1],
+        (key, duplicateKey) -> {
+          throw new IllegalStateException(String.format("Duplicate key %s", key));
```
Comment (dspasojevic) [inline backend/src/main/java/au/com/agiledigital/idea_search/macros/StructuredData.java]: When creating strings, always enclose interpolated values with `[` and `]`.

File: backend/src/main/java/au/com/agiledigital/idea_search/macros/StructuredData.java [hunk 5]
```diff
@@ -1,62 +1,51 @@
...
-//    context.put("formData", currentIdea.getFormData());
+    context.put("data", renderedData);
     return VelocityUtils.getRenderedTemplate("vm/StructuredData.vm", context);
   }
 
+  // converts a string into a capitalised string and splits on capital letters
+  private String headingTransformation(String heading) {
+    String[] headingList = StringUtils.splitByCharacterTypeCamelCase(heading);
+    // to make sure the first letter of the heading is always capitalised
+    headingList[0] = StringUtils.capitalize(headingList[0]);
```
Comment (dspasojevic) [inline backend/src/main/java/au/com/agiledigital/idea_search/macros/StructuredData.java:56]: What if the array is empty?

File: backend/src/main/java/au/com/agiledigital/idea_search/macros/StructuredData.java (hunk 1 above)
Comment (shouvSarker) [reply backend/src/main/java/au/com/agiledigital/idea_search/macros/StructuredData.java]: Updated to assignment on initialisation.

File: backend/src/main/java/au/com/agiledigital/idea_search/macros/StructuredData.java (hunk 2 above)
Comment (shouvSarker) [reply backend/src/main/java/au/com/agiledigital/idea_search/macros/StructuredData.java]: Yes, ran maven:spotless, but it isn't complaining on this particular area. Formatted it a bit to increase readability.

File: backend/src/main/java/au/com/agiledigital/idea_search/macros/StructuredData.java (hunk 3 above)
Comment (shouvSarker) [reply backend/src/main/java/au/com/agiledigital/idea_search/macros/StructuredData.java]: It is not possible to use `Map.entry` as it is an interface, but we can use `AbstractMapSimpleEntry`. Updated with `new AbstractMap.SimpleEntry`

File: backend/src/main/java/au/com/agiledigital/idea_search/macros/StructuredData.java (hunk 5 above)
Comment (shouvSarker) [reply backend/src/main/java/au/com/agiledigital/idea_search/macros/StructuredData.java:56]: Added a check to verify that the array is not empty

File: backend/src/main/java/au/com/agiledigital/idea_search/macros/StructuredData.java (hunk 4 above)
Comment (shouvSarker) [reply backend/src/main/java/au/com/agiledigital/idea_search/macros/StructuredData.java]: Updated

**Pull request merged**